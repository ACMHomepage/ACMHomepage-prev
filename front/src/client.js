import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client';

const link = new HttpLink({
  uri: 'http://localhost:3000/graphql',

  // Use explicit `window.fetch` so tha outgoing requests
  // are captured and deferred until the Service Worker is ready.
  fetch: (...args) => fetch(...args),
});

export const client = new ApolloClient({
  link,
  cache: new InMemoryCache(),
});
