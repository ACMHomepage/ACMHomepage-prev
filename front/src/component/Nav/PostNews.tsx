import { URL as postNewsUrl } from '../../page/PostNews';

import LinkButton from './../LinkButton';

import styles from './styles/PostNews.module.scss';

const PostNews = () => {
  return (
    <LinkButton to={postNewsUrl} className={styles.button}>
      Post News
    </LinkButton>
  );
};

export default PostNews;
