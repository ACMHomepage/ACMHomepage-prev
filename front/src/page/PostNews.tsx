import { Send } from 'lucide-react';
import { useState } from 'react';
import { utilMainPart } from '../config';
import { useCreateNews } from '../store/newsSlice';

import Header from '../component/Header';
import TitleInput from '../component/InputSet/TitleInput';
import ImageInput from '../component/InputSet/ImageInput';
import ContentTextArea from '../component/InputSet/ContextTextArea';

import styles from './styles/PostNews.module.scss';

export const URL = '/postnews';

export default () => {
  const [title, setTitle] = useState('');
  const [content, setContent] = useState('');
  const [imageDataURL, setImageDataURL] = useState('');
  const [createNews, { loading, error, data }] = useCreateNews();

  return (
    <div sx={utilMainPart}>
      <Header.Space>
        <Header>Post News</Header>
      </Header.Space>
      <div className={styles.form}>
        <TitleInput onChange={e => setTitle(e.target.value)} />
        <ImageInput setDataURL={setImageDataURL} />
        <ContentTextArea onChange={e => setContent(e.target.value)} />
        <div className={styles.buttonRow}>
          <button
            className={styles.button}
            onClick={() => createNews(title, imageDataURL, content)}
          >
            <Send size={16} />
            Submit
          </button>
        </div>
      </div>
    </div>
  );
};
