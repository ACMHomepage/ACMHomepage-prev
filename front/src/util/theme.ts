import type {
  Theme,
  ColorModesScale,
  ColorMode,
  ThemeUICSSObject,
  ThemeUIStyleObject,
} from 'theme-ui';
import fill from 'lodash/fill';
import get from 'lodash/get';
import isUndefined from 'lodash/isUndefined';

/******************************************************************************
 * Set the colors.
 *****************************************************************************/
type Color = ColorModesScale[string];
type ColorStyle = ThemeUICSSObject['color'];

interface Colors extends ColorModesScale {
  readonly green: Color;
  readonly gray: Color;
  modes: { dark: ColorMode };
}

const colors: Colors = {
  green: [
    '#F5FBF7', // 0
    '#E8F4EC', // 1
    '#D2E9DB', // 2
    '#B7D8C3', // 3
    '#9EC5AC', // 4
    '#78A288', // 5
    '#5A8068', // 6
    '#446953', // 7
    '#32533F', // 8
    '#244230', // 9
  ],
  gray: [
    '#F9FEFF', // 0
    '#DDE2E3', // 1
    '#C4CBCC', // 2
    '#A9B1B3', // 3
    '#909799', // 4
    '#747B7D', // 5
    '#5D6566', // 6
    '#43494B', // 7
    '#2D3233', // 8
    '#2E3233', // 9
  ],
  modes: {
    dark: {},
  },
};

colors.text = get(colors, 'green.9');
colors.modes.dark.text = get(colors, 'green.2');

colors.background = get(colors, 'green.0');
colors.modes.dark.background = get(colors, 'green.9');

colors.secondaryBackground = get(colors, 'green.2');
colors.modes.dark.secondaryBackground = get(colors, 'green.7');

// Fuck, now those are useless
colors.primary = 'white';
colors.modes.dark.primary = 'white';

colors.secondary = 'white';
colors.modes.dark.secondary = 'white';

interface SetColorConfig {
  setColorToBorderColor: boolean;
}

/**
 * setColor.
 *
 * @param color The text color and border color.
 * @param bg The background's color.
 * @param config Config:
 * - `setColorToBorderColor`: `true`(defalut) | `false`. Set param `color` to
 *   border's color or not.
 */
export const setColor = (
  color: ColorStyle,
  bg: ColorStyle,
  config?: SetColorConfig,
) => {
  if (isUndefined(config)) {
    config = {
      setColorToBorderColor: true,
    };
  }

  return {
    color,
    bg,
    borderColor: config.setColorToBorderColor ? color : undefined,
  };
};

/******************************************************************************
 * Set the break points.
 *****************************************************************************/
const __breakpointNameWithoutInitArray = ['sm', 'md', 'lg', 'xl'] as const;
if ('_' in __breakpointNameWithoutInitArray) {
  throw new Error(
    `Be sure that '_' cannot in the \`__breakpointNameWithoutInitArray\``,
  );
}

export type BreakpointNameWithoutInit =
  typeof __breakpointNameWithoutInitArray[number];
export type BreakpointName = BreakpointNameWithoutInit | '_';
export type BreakpointMapWithoutInit = {
  [size in BreakpointNameWithoutInit]: number;
};
export type BreakpointMap = {
  [size in BreakpointName]: number;
};

// Make breakpoints array and its name map(to breakpoints array's index).
// e.g. makeBreakPoints([['640px', 'sm'], ['780px', 'md']]) === [['640px',
// '780px'], {'sm': 0, 'md': 1}].
const makeBreakPoints = (
  breakpointsAndNames: Array<[string, BreakpointNameWithoutInit]>,
): [Array<string>, BreakpointMapWithoutInit] => {
  if (breakpointsAndNames.length !== __breakpointNameWithoutInitArray.length) {
    throw new Error(
      `Be sure that those names ${__breakpointNameWithoutInitArray},` +
        "are all in the argument `breakpointsAndNames`'s breakpoints' names",
    );
  }

  let result: [Array<string>, { [size: string]: number }] = [[], {}];
  let breakpointNames = new Set();
  for (let i = 0; i < breakpointsAndNames.length; i++) {
    let [size, name] = breakpointsAndNames[i];

    // make sure breakpoint name is unique.
    if (name in breakpointNames) {
      throw new Error("Be sure that breakpoint's name is unique.");
    }
    breakpointNames.add(name);

    result[0].push(size);
    result[1][name] = i;
  }
  return [result[0], result[1] as BreakpointMapWithoutInit];
};

const [breakpoints, breakpointNames] = makeBreakPoints([
  ['640px', 'sm'], // breakpoint 0
  ['780px', 'md'], // breakpoint 1
  ['1024px', 'lg'], // breakpoint 2
  ['1536px', 'xl'], // breakpoint 3
]);

// Make responsive value. (alias: mRV).
// e.g. If breakpointNames(should use `makeBreakPoints` to make) is {'sm': 0,
// 'md': 1, 'lg': 2, 'xl': 3}, then mRV<string>({_: '10px', 'md': '20px'}) ===
// ['10px', null, '20px', null, null].
export const makeResponsiveValue = <T>(
  argument: Partial<{ [size in BreakpointName]: T }>,
): (T | null)[] => {
  let result: (T | null)[] = fill(Array(breakpoints.length + 1), null);
  for (const s in argument) {
    const size = s as BreakpointName;
    const value = argument[size];
    if (isUndefined(value)) continue;
    if (size === '_') {
      result[0] = value;
      continue;
    }
    result[breakpointNames[size] + 1] = value;
  }
  return result;
};

/** Make responsive value. (alias: mRV). */
export const mRV = makeResponsiveValue;

/******************************************************************************
 * Set the radius size for border.
 *****************************************************************************/
const radii = {
  normal: '0.25rem',
  inf: '999999px',
};

interface SetBorderConfig {
  width?: ThemeUICSSObject['borderWidth'];
  radius?: ThemeUICSSObject['borderRadius'];
  style?: ThemeUICSSObject['borderStyle'];
}

/**
 * config:
 * - width: `<WIDTH>` | `'1px'`(defalut).
 * - radius: `<RADII>` | `'normal'`(defalut).
 * - style: `<BORDER-STYLE>` | `'solid'`(default).
 */
export const setBorder = (config?: SetBorderConfig) => {
  if (isUndefined(config)) config = {};
  if (isUndefined(config.width)) config.width = '1px';
  if (isUndefined(config.radius)) config.radius = 'normal';
  if (isUndefined(config.style)) config.style = 'solid';
  return {
    borderWidth: config.width,
    borderRadius: config.radius,
    borderStyle: config.style,
  };
};

/******************************************************************************
 * Set the weights, size for font.
 *****************************************************************************/
const fontWeights: Theme['fontWeights'] = {
  thin: 100,
  extralight: 200,
  light: 300,
  normal: 400,
  medium: 500,
  semibold: 600,
  bold: 700,
  extrabold: 800,
  black: 900,
};

fontWeights.h1 = fontWeights.extrabold;
fontWeights.h2 = fontWeights.bold;
fontWeights.h3 = fontWeights.semibold;
fontWeights.h4 = fontWeights.semibold;
fontWeights.h5 = fontWeights.medium;
fontWeights.h6 = fontWeights.medium;
fontWeights.body = fontWeights.normal;

const fontSizes: Theme['fontSizes'] = {
  xs: '0.75rem',
  sm: '0.875rem',
  base: '1rem',
  lg: '1.125rem',
  xl: '1.25rem',
  '2xl': '1.5rem',
  '3xl': '1.875rem',
  '4xl': '2.25rem',
  '5xl': '3rem',
  '6xl': '3.75rem',
  '7xl': '4.5rem',
  '8xl': '6rem',
  '9xl': '8rem',
};

type FontSizeName = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

export const setFontSize = (name: FontSizeName): ThemeUICSSObject['fontSize'] => {
  switch (name) {
    case 'h1': return mRV({ _: '5xl', md: '6xl', lg: '7xl' });
    case 'h2': return mRV({ _: '4xl', md: '5xl', lg: '6xl' });
    case 'h3': return mRV({ _: '3xl', md: '4xl', lg: '5xl' });
    case 'h4': return mRV({ _: '2xl', md: '3xl', lg: '4xl' });
    case 'h5': return mRV({ _: 'xl', md: '2xl', lg: '3xl' });
    case 'h6': return mRV({ _: 'lg', md: 'xl', lg: '2xl' });
  }
}


/******************************************************************************
 * Set the flexbox.
 *****************************************************************************/
interface SetFlexConfig {
  center?: boolean;
  gap?: ThemeUICSSObject['gap'];
}

/**
 * setFlex.
 *
 * @param config Config the flexbox's style.
 * - `center`: `false`(default) | `true`. Set the content is put at the center
 *   or not.
 * - `gap`: `<LENGTH>` | `undefined`(default).
 */
export const setFlex = ({ center = false, gap }: SetFlexConfig) => {
  return {
    display: 'flex',
    alignItems: center ? 'center' : undefined,
    placeContent: center ? 'center' : undefined,
    gap,
  };
};

/******************************************************************************
 * Default theme.
 *****************************************************************************/
export default {
  config: {
    initialColorModeName: 'light',
  },
  colors,
  breakpoints,
  radii,
  fontWeights,
  fontSizes,
} as Theme;
