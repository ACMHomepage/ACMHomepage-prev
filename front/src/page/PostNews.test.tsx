import PostNews from './PostNews';
import { ThemeProvider } from 'theme-ui';
import { render } from '@testing-library/react';
import { ApolloProvider } from '@apollo/client';
import { BrowserRouter } from 'react-router-dom';

import theme from '../util/theme';
import { client } from '../client';

test('Make sure `PostNews` looks same as before.', () => {
  const { container } = render(
    <ApolloProvider client={client}>
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <PostNews />
        </ThemeProvider>
      </BrowserRouter>
    </ApolloProvider>,
  );

  expect(container).toMatchSnapshot();
});

// TODO: Test its function.
