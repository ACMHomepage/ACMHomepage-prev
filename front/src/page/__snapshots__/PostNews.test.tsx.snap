// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make sure \`PostNews\` looks same as before. 1`] = `
<div>
  <div
    class="css-1cqbbyv"
  >
    <div
      class="css-1gtzog5-HeaderSpace"
    >
      <h1
        class="css-1xvnnad-Header"
      >
        Post News
      </h1>
    </div>
    <div
      class="css-hwxeuh"
    >
      <label
        class="css-f5kwef"
      >
        <svg
          class="css-cyolb1-TitleInput"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 19.5A2.5 2.5 0 016.5 17H20"
          />
          <path
            d="M6.5 2H20v20H6.5A2.5 2.5 0 014 19.5v-15A2.5 2.5 0 016.5 2z"
          />
        </svg>
        <input
          class="css-1yob23c-TitleInput"
          placeholder="Title"
          value=""
        />
      </label>
      <label
        class="css-f5kwef"
      >
        <svg
          class="css-29qq03-ImageInput"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="18"
            rx="2"
            ry="2"
            width="18"
            x="3"
            y="3"
          />
          <circle
            cx="8.5"
            cy="8.5"
            r="1.5"
          />
          <polyline
            points="21 15 16 10 5 21"
          />
        </svg>
        <div
          class="css-1cprmi-Infomation"
        >
          Please upload images...
        </div>
        <input
          class="css-1xn7iig-ImageInput"
          tabindex="-1"
          type="file"
        />
      </label>
      <label
        class="css-f5kwef"
      >
        <svg
          class="css-c1znx3-ContentTextArea"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="21"
            x2="3"
            y1="6"
            y2="6"
          />
          <line
            x1="15"
            x2="3"
            y1="12"
            y2="12"
          />
          <line
            x1="17"
            x2="3"
            y1="18"
            y2="18"
          />
        </svg>
        <textarea
          class="css-sxl6kr-ContentTextArea"
          placeholder="Content"
        />
      </label>
      <div
        class="css-1h1zuwb"
      >
        <button
          class="css-rkdaer"
        >
          <svg
            class="lucide lucide-send"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="22"
              x2="11"
              y1="2"
              y2="13"
            />
            <polygon
              points="22 2 15 22 11 13 2 9 22 2"
            />
          </svg>
          Submit
        </button>
      </div>
    </div>
  </div>
</div>
`;
